examples/example_customdns_test.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) MongoDB, Inc. 2022-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package examples
+
+import (
+	"context"
+	"log"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func resolve(ctx context.Context, cache *dnsCache, in *dns.Conn, out *dns.Conn) {
+	for ctx.Err() == nil {
+		q, err := in.ReadMsg()
+		if err != nil {
+			// TODO: Handle error.
+			log.Fatalf("Unhandled error in ReadMsg: %v", err)
+		}
+		if len(q.Question) != 1 {
+			// Multiple questions in a single query is not actually used in real life.
+			continue
+		}
+
+		a, err := func() (*dns.Msg, error) {
+			cache.lock.Lock()
+			defer cache.lock.Unlock()
+
+			now := time.Now()
+			if rr, ok := cache.records[q.Question[0]]; ok && rr.exp.After(now) {
+				a := new(dns.Msg)
+				a.SetReply(q)
+				a.Compress = false
+				a.Answer = append(a.Answer, rr.record)
+				return a, nil
+			}
+
+			err := out.WriteMsg(q)
+			if err != nil {
+				return nil, err
+			}
+
+			m, err := out.ReadMsg()
+			if err != nil {
+				return nil, err
+			}
+
+			l := len(m.Answer)
+			for i, q := range m.Question {
+				if i >= l {
+					break
+				}
+				a := m.Answer[i]
+				cache.records[q] = &RR{
+					a,
+					now.Add(time.Second * time.Duration(a.Header().Ttl)),
+				}
+			}
+			return m, nil
+		}()
+		if err != nil {
+			// TODO: Handle error.
+			log.Fatalf("Unhandled error in record retrieval: %v", err)
+		}
+
+		if err := in.WriteMsg(a); err != nil {
+			// TODO: Handle error.
+			log.Fatalf("Unhandled error in WriteMsg: %v", err)
+		}
+	}
+}
+
+type RR struct {
+	record dns.RR
+	exp    time.Time
+}
+
+type dnsCache struct {
+	records map[dns.Question]*RR
+	lock    *sync.Mutex
```

Comment:
I use a pointer of mutex mostly because it is an idiom for me... But I think the comment is correct for its metaphor of the difference between a pointer and a concrete struct. Good to learn!